vllm/spec_decode/batch_expansion.py

```diff
@@ -0,0 +1,353 @@
+        proposal tokens, create a new batch where each sequence has a single
+        query token.
+        """
+
+        spec_seqs = [
+            seq_group for seq_group, proposal_len in zip(
+                seq_group_metadata_list, proposal_lens_list)
+            if proposal_len != 0
+        ]
+        spec_indices = [
+            i for i, (_, proposal_len) in enumerate(
+                zip(seq_group_metadata_list, proposal_lens_list))
+            if proposal_len != 0
+        ]
+
+        non_spec_seqs = [
+            seq_group for seq_group, proposal_len in zip(
+                seq_group_metadata_list, proposal_lens_list)
+            if proposal_len == 0
+        ]
+        non_spec_indices = [
+            i for i, (_, proposal_len) in enumerate(
+                zip(seq_group_metadata_list, proposal_lens_list))
+            if proposal_len == 0
+        ]
+
+        target_seq_group_metadata_list = self._create_scoring_model_input(
+            spec_seqs, proposal_token_ids_list)
+        num_scoring_tokens = len(target_seq_group_metadata_list)
+        target_seq_group_metadata_list.extend(non_spec_seqs)
+
+        return spec_indices, non_spec_indices, target_seq_group_metadata_list, num_scoring_tokens
+
+    def _contract_batch(self, original_bs: int,
+                        target_sampler_output: List[SamplerOutput],
+                        proposals: SpeculativeProposals,
+                        num_scoring_tokens: int, non_spec_indices: List[int],
+                        spec_indices: List[int],
+                        k: int) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Contract the expanded batch back into its original size.
+        This maps the scores of speculative tokens back to their original
+        sequences.
+        """
+        (target_token_ids, target_probs, non_spec_target_token_ids,
+         non_spec_target_probs) = self._split_scoring_output(
+             target_sampler_output, num_scoring_tokens)
+
+        # Map distinct sequences used to score each token
+        # of shape [batch_size * k + 1] back to [batch_size, k + 1].
+        batch_size, k = proposals.proposal_token_ids.shape
+
+        target_token_ids = target_token_ids.squeeze().reshape(
+            batch_size, k + 1)
+        target_probs = target_probs.squeeze().reshape(batch_size, k + 1,
+                                                      self._vocab_size)
+
+        all_tokens = torch.ones(
+            original_bs, k + 1, device=self._device, dtype=torch.long) * -1
+        all_probs = torch.zeros(original_bs,
+                                k + 1,
+                                self._vocab_size,
+                                device=self._device,
+                                dtype=torch.float32)
+
+        if non_spec_indices:
+            all_tokens[non_spec_indices, 0] = non_spec_target_token_ids
+            all_probs[non_spec_indices, 1:, :] = non_spec_target_probs
```

Comment:
A bit confused by the 1 here, why starting from 1?

Reply:
This is a bug! Should be `:1`. Saved me some headache during correctness testing, thank you 😄